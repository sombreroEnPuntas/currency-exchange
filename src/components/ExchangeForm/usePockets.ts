import { useDispatch, useSelector } from 'react-redux'

import {
  Currency,
  Pocket,
  getPocketByCurrency,
  setPocketByCurrency,
} from '../../data/pockets'

interface Deps {
  fromTarget: Currency
  toTarget: Currency
}
const usePockets = ({ fromTarget, toTarget }: Deps) => {
  const dispatch = useDispatch()

  const from: Pocket = useSelector(getPocketByCurrency(fromTarget))
  const to: Pocket = useSelector(getPocketByCurrency(toTarget))

  const transaction = (fromValue: string, toValue: string) => {
    dispatch(
      setPocketByCurrency({
        availableAmount: from.availableAmount - parseFloat(fromValue),
        currency: fromTarget,
      })
    )
    dispatch(
      setPocketByCurrency({
        availableAmount: parseFloat(toValue) + to.availableAmount,
        currency: toTarget,
      })
    )
  }

  return { from, to, transaction }
}

export default usePockets
