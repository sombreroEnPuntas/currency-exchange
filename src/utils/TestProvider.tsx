import React from 'react'
import { Provider } from 'react-redux'

import store from '../data'

const TestProvider = ({ children }) => (
  <Provider store={store}>{children}</Provider>
)

export default TestProvider
