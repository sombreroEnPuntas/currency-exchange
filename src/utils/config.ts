export default {
  exchangeRateAPIRoute: 'https://openexchangerates.org/api/latest.json',
  availableCurrencies: ['EUR', 'GBP', 'USD'],
}
