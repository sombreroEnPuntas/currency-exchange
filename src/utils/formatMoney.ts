import { Currency } from '../types'

const roundToTwoDecimals = (number: number) =>
  Math.round((number + Number.EPSILON) * 100) / 100

const localFormatter = (currency?: Currency) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency })

/**
 * Formats a monetary amount as currency when [currency] is given,
 * or two decimals otherwise.
 *
 * @param {number} value monetary amount
 * @param {Currency} [currency] ISO currencies
 */
const formatMoney = (value: number, currency?: Currency) =>
  currency
    ? localFormatter(currency).format(roundToTwoDecimals(value))
    : roundToTwoDecimals(value).toFixed(2)

export default formatMoney
