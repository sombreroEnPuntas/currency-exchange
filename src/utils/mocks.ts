import { Pocket } from '../data/pockets'

export const pocketsMock: Pocket[] = [
  { availableAmount: 58.33, currency: 'GBP' },
  { availableAmount: 25.51, currency: 'USD' },
  { availableAmount: 0, currency: 'EUR' },
]
