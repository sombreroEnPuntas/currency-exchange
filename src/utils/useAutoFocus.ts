import { useRef, useEffect } from 'react'

const useAutoFocus = () => {
  const htmlElRef = useRef(null)

  useEffect(() => {
    htmlElRef.current && htmlElRef.current.focus()
  }, [htmlElRef])

  return htmlElRef
}

export default useAutoFocus
