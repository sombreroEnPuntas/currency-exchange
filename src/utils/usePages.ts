import { useEffect, useState } from 'react'

interface Deps {
  list: any[]
  pageSize: number
}
const usePages = ({ list, pageSize }: Deps) => {
  const [cursor, setCursor] = useState(0)
  const [pages, setPages] = useState([])

  useEffect(() => {
    setPages(
      list.reduce((all, page, index) => {
        const currentPage = Math.floor(index / pageSize)
        all[currentPage] = [].concat(all[currentPage] || [], page)
        return all
      }, [])
    )
  }, [list, pageSize])

  return {
    handleBack: () => setCursor(cursor - 1 < 0 ? pages.length - 1 : cursor - 1),
    handleNext: () => setCursor(cursor + 1 > pages.length - 1 ? 0 : cursor + 1),
    page: pages[cursor] || [],
    pageNumber: cursor + 1,
    totalPages: pages.length,
  }
}

export default usePages
