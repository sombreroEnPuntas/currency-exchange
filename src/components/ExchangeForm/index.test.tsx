import React from 'react'
import { fireEvent, render } from '@testing-library/react'

// Tested Unit
import Component from '.'

// Deps
import { useRouter } from 'next/router'
import usePockets from './usePockets'

// Utils
import TestProvider from '../../utils/TestProvider'
import { pocketsMock } from '../../utils/mocks'

// Mocks
jest.mock('next/router')
jest.mock('./usePockets')
const testFromPocket = pocketsMock.find(({ currency }) => currency === 'GBP')
const testToPocket = pocketsMock.find(({ currency }) => currency === 'USD')

// Mock data
const setMock = ({ from, to, transaction, push }) => {
  ;(useRouter as jest.Mock).mockImplementation(() => ({
    push,
  }))
  ;(usePockets as jest.Mock).mockImplementation(() => ({
    from,
    to,
    transaction,
  }))
}

const setup = ({
  from = testFromPocket,
  to = testToPocket,
  transaction = jest.fn(),
  push = jest.fn(),
} = {}) => {
  setMock({ from, to, transaction, push })

  const utils = render(
    <TestProvider>
      <Component />
    </TestProvider>
  )

  const inputEl = utils.getByLabelText('GBP')
  const convertedEl = utils.getByLabelText('USD')
  const cancelCTA = utils.getByText(/Cancel/)
  const exchangeCTA = utils.getByText(/Exchange/)

  return { inputEl, convertedEl, cancelCTA, exchangeCTA, ...utils }
}

describe('ExchangeForm', () => {
  it('Renders default state', () => {
    const { container, inputEl, convertedEl, cancelCTA, exchangeCTA } = setup()

    expect(cancelCTA).toBeVisible()
    expect(exchangeCTA).toBeVisible()
    expect(inputEl).toBeVisible()
    expect(inputEl).toHaveFocus()
    expect(convertedEl).toBeVisible()

    expect(container).toMatchSnapshot()
  })

  it('Calculates exchange during change', () => {
    const { inputEl, convertedEl } = setup()
    const userInputValue = '23.50'
    const convertedValue = 31.59

    fireEvent.change(inputEl, { target: { value: userInputValue } })

    expect(convertedEl).toHaveValue(convertedValue)
  })

  it('Validates input', () => {
    const { inputEl } = setup()

    fireEvent.change(inputEl, { target: { value: '-1' } }) // min

    expect(inputEl).toBeInvalid()

    fireEvent.change(inputEl, { target: { value: '58.34' } }) // max

    expect(inputEl).toBeInvalid()

    fireEvent.change(inputEl, { target: { value: '58.33' } }) // valid

    expect(inputEl).toBeValid()
  })

  it('Goes back to home on Cancel', () => {
    const push = jest.fn()
    const { cancelCTA } = setup({ push })

    fireEvent.click(cancelCTA)

    expect(push).toHaveBeenCalledWith('/')
  })

  it('Executes a Transaction on Exchange', () => {
    const transaction = jest.fn()
    const { inputEl, exchangeCTA } = setup({ transaction })
    const userInputValue = '23.50'
    const convertedValue = '31.59'

    fireEvent.change(inputEl, { target: { value: userInputValue } })

    fireEvent.click(exchangeCTA)

    expect(transaction).toHaveBeenCalledWith(userInputValue, convertedValue)
  })
})
