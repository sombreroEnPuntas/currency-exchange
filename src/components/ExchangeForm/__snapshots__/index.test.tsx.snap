// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeForm Renders default state 1`] = `
<div>
  <form
    id="exchange"
    name="exchange"
  >
    <label
      for="from"
    >
      GBP
    </label>
    <input
      id="from"
      max="58.33"
      min="0"
      name="from"
      required=""
      step="0.01"
      type="number"
      value=""
    />
    <div>
      You have Â£58.33
    </div>
    <label
      for="to"
    >
      USD
    </label>
    <input
      id="to"
      max="25.51"
      min="0"
      name="to"
      readonly=""
      step="0.01"
      tabindex="-1"
      type="number"
      value=""
    />
    <div>
      You have $25.51
    </div>
    <button
      type="button"
    >
      Cancel
    </button>
    <button
      type="submit"
    >
      Exchange
    </button>
  </form>
</div>
`;
